### PR TITLE
Vendor installer release-4.10-azure

### PR DESCRIPTION
**Pull request by darthhexx**
### Which issue this PR addresses:

Switching to OCP 4.10 for cluster installations.
https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/12859516/

### What this PR does / why we need it:

Re-vendors and makes appropriate changes to change to the OpenShift Installer release-4.10. Most notably the move of the AzureMachineProviderSpec definition to the OpenShift API and subsequent changes to the machine set specifications.

Reference commit: https://github.com/openshift/installer/commit/f9725ddd9496b999ac06bd0343791d7c50b00d48

### Note

* This PR currently points to my installer fork: https://github.com/darthhexx/installer-aro/tree/release-4.10-azure This will be changed and the *blocked* label removed, once this PR https://github.com/jewzaam/installer-aro/pull/12 is merged, so that reviewers are able to create 4.10 clusters in the meantime.

* The PR contains a work around for unmarshalling existing AzureMachineProviderSpec without importing sigs.k8s.io/cluster-api-provider-azure to simplify vendoring that I'm specifically looking for reviews on: **pkg/operator/controllers/machine/machine.go#L42-L63**

### Test plan for issue:

Existing unit and e2e tests.

### Is there any documentation that needs to be updated for this PR?

The repository documentation has been changed in the PR and a release to https://github.com/Azure/OpenShift/releases will need to be posted.

### CODE REVIEW

**Comment by bennerv** [comment]
Looks like we'll need some new CI VMs with go 1.17 on them for this as well.  Do you think you can add these on your own @darthhexx or do you need some help?

**Comment by darthhexx** [comment]
@bennerv I know the changes need to be made in `pkg/deploy/generator/resources_dev.go` and followed by a `make generate`, but I do have some questions:
1. If I use the `az deployment` command in `deploy_env_dev_ci` to deploy a go 1.17 capable VMSS, will the VMs be _included_ in the ARO CI Agent Pool or will they effectively replace the existing set?
2. There is no go-toolset-1.17 currently available in RHEL 8. Should we use the precompiled binaries from go.dev, an RPM from elsewhere, or something else?
3. I see no documentation or code around the setting of the VM test capabilities. Is this step done manually after they're deployed?

**Comment by bennerv** [comment]
> @bennerv I know the changes need to be made in `pkg/deploy/generator/resources_dev.go` and followed by a `make generate`, but I do have some questions:
> 
>     1. If I use the `az deployment` command in `deploy_env_dev_ci` to deploy a go 1.17 capable VMSS, using a different VMS name, will the VMs be _included_ in the ARO CI Agent Pool or will they effectively replace the existing set?

It uses an ARM template which just looks for ideal state.  Since the names of the VMSS will be the same, it won't actually do anything.  Before deploying, you'll have to update all the names in `resources_dev.go` to something like `ci-vmss-1.17`. 


>     2. There is no go-toolset-1.17 currently available in RHEL 8. Should we use the precompiled binaries from go.dev, an RPM from elsewhere, or something else?


In rhel8 +, they changed the package so it's no longer named `go-toolset-1.17`, but rather publish multiple versions for the package `go-toolset`.  You can see that there's a 1.17 go toolset for el8 [here](https://access.redhat.com/downloads/content/go-toolset/1.17.7-1.module+el8.6.0+14297+32a15e19/x86_64/fd431d51/package)

Should be `go-toolset-1.17.7-1.module+el8.6.0+14297+32a15e19` though

>     3. I see no documentation or code around the setting of the VM test capabilities. Is this step done manually after they're deployed?

This is also located in `resources_dev.go` as it's set in a configuration file on deploy.  Check [this](https://github.com/Azure/ARO-RP/blob/master/pkg/deploy/generator/resources_dev.go#L354) and update it to `go-1.17` for the newest and run a  `make generate`.

I'm going to make sure the ADO token we use to deploy the CI VMs in our INT secrets is up-to-date.  If you run into any issues here, please let me know.  

We have stories to improve this process, but never picked them up at the time due to 1ESHosted changes and we thought EV2 may impact these as well, but they won't . Right now it's just tribal knowledge (which we had to learn when the people who set this up left the team as well).

File: hack/validate-imports/validate-imports.go
Comment (bennerv) [inline hack/validate-imports/validate-imports.go:157]: These are covered under a different case now that they're in `client-go`. You can remove all the updated case statements now.    

see [this](https://github.com/Azure/ARO-RP/pull/2070/files#diff-8f4e9c2f8faf0eaf4cad0824e4d78b7a1c5177a4fa39d885d1c19cb0ec625a2eL90-L98) for details.  You already updated the imports everywhere so should be good once you remove these.    

```suggestion
```

File: pkg/operator/controllers/machine/machine.go
Comment (bennerv) [inline pkg/operator/controllers/machine/machine.go:62]: Why do we need both "unmarshals" here?  Is it to support the older types as well as the newer types?

**Review by bennerv — CHANGES_REQUESTED** [review]
Great work!  Couple nits to address.  Will try to create a cluster sometime next week.

File: Dockerfile.aro-multistage
Comment (darthhexx) [reply Dockerfile.aro-multistage:7]: @bennerv there are none returned from `podman image search --list-tags registry.access.redhat.com/ubi{7,8}/go-toolset` so I went with the CI version that's used for testing OpenShift 4.10.

I actually suggest we change to building from the latest MSFT Go FIPS source so that going forward we aren't dependant on other parties for our base image. @cadenmarchese I have recently done this to get 1.17 support for our CI Agent Pool, but it sounds like you are already working on solving this with CDPx builds?

File: hack/validate-imports/validate-imports.go
Comment (darthhexx) [reply hack/validate-imports/validate-imports.go:157]: Thanks for spotting that :+1: I'll make the change in my next rebase.

File: pkg/operator/controllers/machine/machine.go
Comment (darthhexx) [reply pkg/operator/controllers/machine/machine.go:62]: For previous versions on OCP the **AzureMachineProviderSpec** is defined with `azureproviderconfig.openshift.io/v1beta1` but that has been moved to `machine.openshift.io/v1beta1` in 4.10:
https://github.com/openshift/installer/commit/f9725ddd9496b999ac06bd0343791d7c50b00d48#diff-54168990f7f297f463e06a145f714bd63582499b41b4be695f49f5f16dc6f76eR110

In this release we are moving to the new API, but we also need to be able to verify machines in all of the existing machinesets in the fleet, which are defined under `azureproviderconfig.openshift.io`. To simplify the switchover the raw string is checked for the old format and that is marshalled into the "new" `machineapi.AzureMachineProviderSpec` struct, which has exactly the same definition.

Note that the `machineapi.AzureMachineProviderSpec` is also not registered by default, so there is also this related change to ensure the operator machine validation still works:
https://github.com/darthhexx/ARO-RP/blob/2ba04295d89f14472c909530c262fda3e088d91e/pkg/util/scheme/scheme.go#L43-L44

File: hack/validate-imports/validate-imports.go
Comment (darthhexx) [reply hack/validate-imports/validate-imports.go]: These have been removed.

**Review by ross-bryan — DISMISSED** [review]
lgtm. a couple of comments, and a change we can make in a follow up PR

File: hack/update-go-module-dependencies.sh
Comment (darthhexx) [reply hack/update-go-module-dependencies.sh:63]: This was temporary, while waiting for the installer PR to be merged. now that it is I am going to rebase & re-vendor.

File: cmd/aro/ensurebuildtag_aro.go
Comment (darthhexx) [reply cmd/aro/ensurebuildtag_aro.go:1]: It's a go 1.17 requirement.

File: test/e2e/e2e_test.go
Comment (darthhexx) [reply test/e2e/e2e_test.go:1]: Yip, auto-added in go 1.17.

**Review by ross-bryan — APPROVED** [review]
lgtm

File: Dockerfile.aro-multistage
Comment (hawkowl) [reply Dockerfile.aro-multistage:7]: ubi8 now has 1.17 released: https://catalog.redhat.com/software/containers/ubi8/go-toolset/5ce8713aac3db925c03774d1

**Comment by hawkowl** [comment]
@darthhexx Apart from the ubi8/go-toolset:1.17 image thing, this is a thumbs up from me.

**Comment by darthhexx** [comment]
@hawkowl the plan is to create a new PR after this one is merged to switch to using the ubi8 go-toolset image, along with some re-vendoring.

**Pull request merged**
